docs: comparison with helix and apollo-server (#1586)
dotansimha/graphql-yoga: website/v3/docs/comparison.mdx

---
id: comparison
title: Comparison with other JavaScript GraphQL server libraries
sidebar_label: Comparison
---

## GraphQL Yoga and Apollo Server

This comparison highlights the main differences when using Yoga instead of Apollo Server.

### No Framework-specific Packages

> One of the goals of graphql-yoga was to have as few `graphql-yoga` packages as possible.

Reviewer: ```suggestion
One of the goals of `graphql-yoga` is to have as few packages as possible.
```

Yoga is designed around the W3C Request/Response specification (often just referred to as ["Fetch API"](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API)),
which is already adopted by deno, cloudflare workers, and browsers.

For Node.js environments and frameworks (which unfortunately do not have full W3C Request/Response support), graphql-yoga comes with a small compatibility layer for making it work.

You only need to search for [`apollo-server-` on NPM](https://www.npmjs.com/search?q=apollo-server-) to see how many variants and packages there are for apollo and how hard it is to maintain.

### Run anywhere

> As mentioned in the above paragraph graphql-yoga is built around W3C Request/Response which is supported in all major serverless/worker/"serverful" JavaScript environments.

Reviewer: ```suggestion
As mentioned in the above paragraph `graphql-yoga` is built around W3C Request/Response which is supported in all major serverless/worker/"serverful" JavaScript environments.
```

Apollo Server plans to drop active support for cloudflare workers and [pushes it onto the community](https://github.com/apollographql/apollo-server/issues/6034).
GraphQL Yoga will continuously support all platforms and runtimes and
Furthermore, Yoga has a full end2end testing suite that actually deploys to all those runtimes in order to ensure integrity and prevent unexpected issues.

### Powerful Plugin System

> Envelopis a powerful GraphQL plugin system, that is unfortunately not fully compatible with Apollo Server.

Reviewer: ```suggestion
Envelop’s powerful GraphQL plugin system, that is unfortunately not fully compatible with Apollo Server.
```

- [Dependency visualizer graphql-yoga](https://npm.anvaka.com/#/view/2d/graphql-yoga)
- [Dependency visualizer apollo-server-micro](https://npm.anvaka.com/#/view/2d/apollo-server-micro)

GraphQL Yogas APIs are designed for code-splitting as supported by popular bundlers.

## GraphQL Yoga and GraphQL Helix

We are maintainers of both libraries today.

The Guild initially adopted, improved, and recommended graphql-helix.

While working on it, we realized that it to some extent allowed building more extensible GraphQL server set-ups, but on the other hand, required boilerplate code and isn't easily adoptable by new GraphQL developers.
We also tried adding support for other environments such as cloudflare workers, but adding it as an after-thought wasn't that straightforward.
Eventually, after a lot of trying we figured that the approach of adding it as an after-thought is the actual limitation, and instead of the other way around (W3C Request/Response first, Node.js HTTP Response/Request as a light-weight layer on top) might be a better solution.

This put us in a difficult position.

> The Guild does not own graphql-helix and we did not want to do drastic changes to it, without verifying that it actually works.

Reviewer: ```suggestion
The Guild does not own `graphql-helix` and we did not want to do drastic changes to it, without verifying that it actually works.
```

> At the same time we were already building graphql-yoga based on helix. So we decided to actually do this experiment within graphql-yoga and remove helix as a dependency from yoga.

Reviewer: ```suggestion
At the same time we were already building `graphql-yoga` based on helix. So we decided to actually do this experiment within graphql-yoga and remove helix as a dependency from yoga.
```

Today we can happily say that it worked out fine!

> Today we recommend graphql-yoga over graphql-helix, if you want a batteries-included solution for GraphQL server development.

Reviewer: ```suggestion
Today we recommend `graphql-yoga` over `graphql-helix`, if you want a batteries-included solution for GraphQL server development.
```

> GraphQL Yoga is easier to get started, has nicer defaults, and requires less boilerplate while giving you the same extensibility as graphql-helix!

Reviewer: ```suggestion
GraphQL Yoga is easier to get started, has nicer defaults, and requires less boilerplate while giving you the same extensibility as `graphql-helix`!
```
